### PR TITLE
T/208: TextTransformation should only consider typing changes

### LINKED ISSUE #208: Text transformation happens on backspace
Since text transformation does not check whether actually typing has happened, text transformation triggers on backspace.

Jul-08-2019 15-22-11

### LINKED ISSUE #214: Introduce a way to check if changes were introduced by typing
I think it would be useful to have a way to check if given changes were done through typing. I see it as a method on `Typing` or `InputCommand` instance.

This could be solved in one of two ways:

1. Add special batch type (yeah, we made precedence with undo :P).
2. Store all `InputCommand` batches in some kind of a `WeakSet`.

Then, we could use it like this:

```js
modelDocument.on( 'change', ( evt, batch ) => {
    if ( typingPlugin.isTyping( batch ) ) {
        // ...
    }
} );
```

```js
modelDocument.registerPostFixer( writer => {
    if ( typingPlugin.isTyping( writer.batch ) ) {
        // ...
    }
} );
```

When this is done, we should fix changes introduced by this PR: https://github.com/ckeditor/ckeditor5-typing/pull/213/files

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduce `Input.isInput()` method. Closes #214. Also introduces a fix for the TextTransformation feature - it willl only consider typing changes. Closes #208.

---

### Additional information

* This PR fixes two bugs with deletion: transformation on delete and transformation on merging (ie paragraphs).
* This PR also introduces `Input.isInput()` check. #214.

### CODE REVIEW

**Comment by scofalik** [comment]
Generally, I dislike heuristics like the one that checks if typing happened. For now, I am okay merging this, but I will make a follow-up in typing to solve the problem of recognizing if typing happened. This is not the first case where this pops up and where it would be useful to have this kind of check.

**Comment by scofalik** [comment]
https://github.com/ckeditor/ckeditor5-typing/issues/214

**Comment by jodator** [comment]
@scofalik I had other heuristics that check if the change wasn't `delete` nor `merge`. Both included checking the operations in the batch. After checking other scenarios I've just used the one from autoformat...

**Comment by scofalik** [comment]
I understand but still, it's better to be sure. If we can check something for sure with low cost, I'd rather have that than a heuristic, so I've made a follow-up.

**Comment by jodator** [comment]
So I'll fix this with #214 as this is much neede IMO :)

**Comment by jodator** [comment]
@scofalik please re-review this with `Input.isInput()` added.

File: src/input.js [hunk 1]
```diff
@@ -40,4 +40,19 @@ export default class Input extends Plugin {
 		injectUnsafeKeystrokesHandling( editor );
 		injectTypingMutationsHandling( editor );
 	}
+
+	/**
+	 * Checks batch if it is a result of user input - e.g. typing.
+	 *
+	 * **Note:** This method checks if the batch was created using {@link module:typing/inputcommand~InputCommand 'input'}
+	 * command as typing changes coming from user input are inserted to the document using that command.
+	 *
```
Comment (scofalik) [inline src/input.js]: I think it would be nice to provide an example here. Check the issue discussion -- I've posted an example snippet there.

Comment (scofalik) [reply src/input.js]: It doesn't cost us much and if anyone would wonder how to exactly use it then they will have a snippet.

File: src/inputcommand.js [hunk 2]
```diff
@@ -35,6 +35,15 @@ export default class InputCommand extends Command {
 		 * @member {module:typing/utils/changebuffer~ChangeBuffer} #_buffer
 		 */
 		this._buffer = new ChangeBuffer( editor.model, undoStepSize );
+
+		/**
+		 * Stores batches created by the input command. The batches are used to differentiate input batches from other batches using
+		 * {@link module:typing/input~Input#isInput} method.
+		 *
+		 * @type {WeakSet<module:engine/model/batch~Batch>}
+		 * @protected
```
Comment (scofalik) [inline src/inputcommand.js]: Does it have to be `protected`? Why not `private`?

File: src/textwatcher.js [hunk 3]
```diff
@@ -79,8 +79,9 @@ export default class TextWatcher {
 	 *
 	 * @private
 	 * @param {'data'|'selection'} suffix Suffix used for generating event name.
+	 * @param {Object} data Data object for event.
 	 */
-	_evaluateTextBeforeSelection( suffix ) {
+	_evaluateTextBeforeSelection( suffix, data = {} ) {
```
Comment (scofalik) [inline src/textwatcher.js]: Do we still need this change?

File: src/inputcommand.js (hunk 2 above)
Comment (jodator) [reply src/inputcommand.js]: It's used by other class (`Indent`) so this must be protected not private.

File: src/textwatcher.js (hunk 3 above)
Comment (jodator) [reply src/textwatcher.js]: I'll check.

File: src/inputcommand.js (hunk 2 above)
Comment (scofalik) [reply src/inputcommand.js]: Okay, I didn't notice these are in two different classes.

File: src/textwatcher.js (hunk 3 above)
Comment (jodator) [reply src/textwatcher.js]: We do pass the `batch` for the `matched:data` event.

File: src/input.js (hunk 1 above)
Comment (jodator) [reply src/input.js]: I've added a really simple one - probably is enough.

File: src/input.js [hunk 4]
```diff
@@ -40,4 +40,27 @@ export default class Input extends Plugin {
...
 		injectTypingMutationsHandling( editor );
 	}
+
+	/**
+	 * Checks batch if it is a result of user input - e.g. typing.
+	 *
+	 *		const input = editor.plugins.get( 'Input' );
+	 *
+	 *		editor.model.document.on( 'change:data', ( evt, batch ) => {
+	 *			if ( input.isTyping( batch) ) {
```
Comment (scofalik) [inline src/input.js]:
```suggestion
	 *			if ( input.isTyping( batch ) ) {
```

Comment (jodator) [reply src/input.js]: Thanks!

**Pull request merged**